### ISSUE TITLE
Buffer overflow on too long output directory name

### ISSUE THREAD

**Issue by haxxor1338** [twogood/unshield#58]
When trying to extract a cab file using a too long custom output path, the program crashes because of a buffer overflow. A contrived example call is:
`unshield -d $(python -c "print 'A'*300") x ./abc/_SYS1.CAB`

The line in question is here:
https://github.com/twogood/unshield/blob/master/src/unshield.c#L388
where `char dirname[256];` is fixed-length, but `output_directory`'s length is (unbound) variable, coming directly from the argument.

To fix this, I suggest using `strncpy` or checking that the `output_directory` doesn't exceed some length.

Below is the output it gives me:
```
Cabinet: ./abc/_SYS1.CAB
*** buffer overflow detected ***: unshield terminated
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7fb29fea97e5]
/lib/x86_64-linux-gnu/libc.so.6(__fortify_fail+0x5c)[0x7fb29ff4a56c]
/lib/x86_64-linux-gnu/libc.so.6(+0x116570)[0x7fb29ff48570]
/lib/x86_64-linux-gnu/libc.so.6(+0x1158c2)[0x7fb29ff478c2]
unshield[0x4019d9]
unshield[0x40219d]
unshield[0x401f26]
unshield[0x401693]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf0)[0x7fb29fe52830]
unshield[0x4017b0]
======= Memory map: ========
[...]
```

**Comment by twogood** [comment]
Nice catch!

PS. Nice username! ;-)

**Comment by twogood** [comment]
I can't reproduce this on the latest version of Unshield.

**Comment by haxxor1338** [comment]
My apologies! I must have used the version of unshield provided by Ubuntu by accident (`apt install unshield`), since I compiled it manually and called it incorrectly afterwards.

This time I made sure I used the version from this source, and can still get an overflow, granted I am not 100% sure yet where exactly the problem is.

I hope it's okay to paste the debug output in the issue here:

```
ubuntu@ubuntu16:~/unshield$ /var/tmp/unshield/bin/unshield -D 3 -d $(python -c "print 'A'*1000") x ./abc/_SYS1.CAB
[unshield_fopen_for_reading:61] File ./abc/_SYS1.hdr not found even case insensitive

[unshield_read_headers:233] Could not open .hdr file 1. Reading header from .cab file 1 instead.
[unshield_read_headers:296] Version 0x01000004 handled as major version 0
[unshield_get_cab_descriptor:82] Cabinet descriptor: 00001324 000000ee 000000ee 00000004
[unshield_get_cab_descriptor:84] Directory count: 1
[unshield_get_cab_descriptor:85] File count: 4
[unshield_file_group_new:15] File group descriptor offset: 00000b36
[unshield_file_group_new:30] File group 00000b36 first file = 0, last file = 65536
[unshield_file_group_new:15] File group descriptor offset: 00000b93
[unshield_file_group_new:30] File group 00000b93 first file = 0, last file = 65536
[unshield_file_group_new:15] File group descriptor offset: 00000bf0
[unshield_file_group_new:30] File group 00000bf0 first file = 0, last file = 65536
[unshield_fopen_for_reading:61] File ./abc/_SYS2.hdr not found even case insensitive

[unshield_read_headers:233] Could not open .hdr file 2. Reading header from .cab file 2 instead.
[unshield_fopen_for_reading:61] File ./abc/_SYS2.cab not found even case insensitive

Cabinet: ./abc/_SYS1.CAB
[unshield_read_file_descriptor:47] File descriptor offset 0: 00001538
[unshield_read_file_descriptor:63] Name offset:      000000bc
[unshield_read_file_descriptor:64] Directory index:  00000000
[unshield_read_file_descriptor:65] Flags:            0004
[unshield_read_file_descriptor:66] Expanded size:    0000548a
[unshield_read_file_descriptor:67] Compressed size:  000011cb
[unshield_read_file_descriptor:68] Data offset:      0000161a
Failed to create directory AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/SYSFG_0009I32/


Extraction failed.
Possible directory traversal attack for: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
To be placed at: /home/ubuntu/unshield/AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA

Segmentation fault (core dumped)
```

**Comment by twogood** [comment]
Same for me with 1000 characters in path!

**Comment by maqifrnswa** [comment]
Ubuntu compiles with:
```
> dpkg-buildflags --dump
CFLAGS=-g -O2 -fdebug-prefix-map=/home/showard=. -fstack-protector-strong -Wformat -Werror=format-security
CPPFLAGS=-Wdate-time -D_FORTIFY_SOURCE=2
...
```
which is why the buffer overflow detection showed up in the apt-get unshield but the buffer overflow is not detected in the one compiled without those flags. It looks like output_directory overflows but is not detected, but is caught by the directory traversal check (since that buffer is meaningless, there's no reason it will match the expected output and therefore appears to be a directory traversal attack).
strncopy up to [256-1] would work, or malloc the path_max and strncopy up to path_max-1 (see https://github.com/twogood/unshield/blob/master/src/unshield.c#L379)

**Comment by twogood** [comment]
Due to personal priorities, I need a PR for this even if it's an easy fix... :-(

**Comment by maqifrnswa** [comment]
I linked an in-progress branch. So far I fixed a bunch of potential buffer overflows. Now the problem is that we're only checking for the total path length to be < PATH_MAX, and not individual path components to be < NAME_MAX. so ```mkdir``` fails if ```directory``` is longer than 255 characters, even if the total path is shorter.
Todo:

- [ ] my fixes are crude, can be made much cleaner. For example, a bunch of my checks before ```strcat``` calls can  simply be changed to 
```
if(strlen(strncat(dirname, string_to_cat, path_max))>=path_max)
 {throw the error}
```

- [ ] need to check that ```strlen(directory) < NAME_MAX```, maybe only check after mkdir fails (https://github.com/twogood/unshield/blob/7b88187ea36d36907b501f0828dc40586703ab75/src/unshield.c#L117) Will need add the following so it works on windows.
```
+#ifndef NAME_MAX
+#define NAME_MAX FILENAME_MAX
+#endif
```
